### PR TITLE
feat: Add cosine similarity query

### LINKED ISSUE #3349: Compute similarity between two vectors

### PR DESCRIPTION
**Pull request by fredcarle**
## Relevant issue(s)

Resolves #3349 

## Description

This PR adds the possibility to calculate the cosine similarity between a vector field and a given vector.

To achieve this we added the `_similarity` system field which take a target field (part of the parent object) and vector as parameter.
```gql
query {
  User{
    _similarity(pointsList: {vector: [1, 2, 0]})
  }
}
```

Note that the added code to mapper and planner is more of a "bolt on" addition given the current state of that part of the code base. A refactor is expected in the future.

Future work will allow giving a `content` parameter instead of the `vector` if the target field has embedding generation configured. This will enable out-of-the-box RAG queries.

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

Added cosine similarity to integration testing

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: client/request/similarity.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package request
+
+// Similarity is a functional field that defines the
+// parameters to calculate the cosine similarity between two vectors.
```
Comment (AndrewSisley) [inline client/request/similarity.go:14]: suggestion: It might be good to note that they must be the same length.

File: internal/planner/select.go [hunk 2]
```diff
@@ -61,6 +61,8 @@ type selectTopNode struct {
 	// selectNode is used pre-wiring of the plan (before expansion and all).
 	selectNode *selectNode
 
+	similarity []*similarityNode
```
Comment (AndrewSisley) [inline internal/planner/select.go:64]: todo: Please add a comment (preferably linked to a follow-up ticket), noting what this is, and immediate future plans.

File: internal/request/graphql/parser/query.go [hunk 3]
```diff
@@ -223,6 +223,31 @@ func parseAggregate(
...
+	field *ast.Field,
+) (*request.Similarity, error) {
+	fieldDef := gql.GetFieldDef(exe.Schema, parent, field.Name.Value)
+	arguments := gql.GetArgumentValues(fieldDef.Args, field.Arguments, exe.VariableValues)
+	var target string
+	var vector any
+	for _, argument := range field.Arguments {
+		target = argument.Name.Value
+		v := arguments[target].(map[string]any)
+		vector = v["vector"]
```
Comment (AndrewSisley) [inline internal/request/graphql/parser/query.go:238]: todo: Please create a const for `"vector"` in the correct file (IIRC there is one in the graphql package or one of it's child-packages)

File: internal/request/graphql/schema/generate.go [hunk 4]
```diff
@@ -846,6 +856,40 @@ func (g *Generator) genAverageFieldConfig(obj *gql.Object) (gql.Field, error) {
 	return field, nil
 }
 
+func (g *Generator) genSimilarityFieldConfig(obj *gql.Object) (gql.Field, error) {
+	field := gql.Field{
+		Name:        request.SimilarityFieldName,
+		Description: "Returns the cosine similarity between the specified field and the provided vector.",
```
Comment (islamaliev) [inline internal/request/graphql/schema/generate.go:862]: suggestion: I think the description should be defined in `schemaTypes` package

File: internal/request/graphql/schema/generate.go [hunk 5]
```diff
@@ -846,6 +856,40 @@ func (g *Generator) genAverageFieldConfig(obj *gql.Object) (gql.Field, error) {
...
+			continue
+		}
+
+		inputObject := gql.NewInputObject(gql.InputObjectConfig{
+			Name:        genSimilaritySelectorName(obj.Name(), objectField.Name),
+			Description: objectField.Description,
+			Fields: gql.InputObjectConfigFieldMap{
+				"vector": &gql.InputObjectFieldConfig{
+					Type:        gql.NewNonNull(gql.NewList(listType.OfType)),
+					Description: "A vector of the same type as the field to compute the cosine similarity with.",
```
Comment (islamaliev) [inline internal/request/graphql/schema/generate.go:879]: suggestion: I think this also should be extracted.

File: tests/integration/query/simple/with_similarity_test.go [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+				},
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+// This test documents a bug where having two aliases in a logical _or operator
+// return no results even though in the tests bellow 1 should be returned.
```
Comment (islamaliev) [inline tests/integration/query/simple/with_similarity_test.go:337]: question: is there a ticket to tackle the bug? If yes, I would suggest to link it here.

File: tests/integration/query/simple/with_similarity_test.go [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+					name: String
+					pointsList: [Int!]
+				}`,
+			},
+			testUtils.CreateDoc{
+				DocMap: map[string]any{
+					"name":       "John",
+					"pointsList": []int64{2, 4, 1},
+				},
+			},
```
Comment (islamaliev) [inline tests/integration/query/simple/with_similarity_test.go:131]: todo: all your tests have well-defined types, but I think it's also important with `[]any`.

I'm not 100% sure, but `[]any` should be the type if instead of `DocMap` we use `Doc`:
```go
testUtils.CreateDoc{
	Doc: `{
		"name":       "John",
		"pointsList": [2, 4, 1],
	}`,
},
```

File: tests/integration/query/simple/with_similarity_test.go
Comment (islamaliev) [inline tests/integration/query/simple/with_similarity_test.go:1]: todo: please add a test that tries to use similarity on non-numeric array, like `[String!]`

**Review by islamaliev — COMMENTED** [review]
Overall looks good (ignoring the hacky part).

I have mostly doc suggestions and few test requests.

File: client/request/similarity.go (hunk 1 above)
Comment (fredcarle) [reply client/request/similarity.go:14]: done

File: internal/request/graphql/parser/query.go (hunk 3 above)
Comment (fredcarle) [reply internal/request/graphql/parser/query.go]: done

File: internal/planner/mapper/similarity.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+	*core.DocumentMapping
+
+	// The targetted field for the cosine similarity
+	SimilarityTarget Targetable
+
+	// The vector to compare the target field to.
+	Vector any
+}
+
+func (a *Similarity) CloneTo(index int) Requestable {
```
[the remark this answers is not in the document]
Comment (fredcarle) [reply internal/planner/mapper/similarity.go]: These might not have been useful. They were copy pasted. I removed them and the tests all seem to pass without issue so it would have been dead code.

File: internal/request/graphql/schema/generate.go (hunk 4 above)
Comment (fredcarle) [reply internal/request/graphql/schema/generate.go:862]: We discussed not long ago that the consensus for description like these was to have them inline and that we would actually work towards returning the extracted ones back to inline.

File: internal/request/graphql/schema/generate.go (hunk 5 above)
Comment (fredcarle) [reply internal/request/graphql/schema/generate.go:879]: Same as above :)

**Review by islamaliev — APPROVED** [review]
Looks great. Thanks for quick adjustments

File: tests/integration/query/simple/with_similarity_test.go (hunk 6 above)
Comment (fredcarle) [reply tests/integration/query/simple/with_similarity_test.go:431]: Added an issue and linked it here :)

File: tests/integration/query/simple/with_similarity_test.go
Comment (fredcarle) [reply tests/integration/query/simple/with_similarity_test.go:1]: done. Thanks for flagging the different length one. I had forgotten about ensuring they were the same length.

File: tests/integration/schema/default_fields.go [hunk 9]
```diff
@@ -149,6 +150,14 @@ var aggregateFields = fields{
...
 }
 
+var similarityField = Field{
+	"name": "_similarity",
+	"type": map[string]any{
+		"kind": "SCALAR",
+		"name": "Float",
+	},
+}
+
```
Comment (fredcarle) [inline tests/integration/schema/default_fields.go:160]: > todo: Please make sure you add an introspection test(s) to make sure the new input type shows up in the correct shape in the correct place without breaking the GQL spec.

@AndrewSisley I added this here to ensure it's part of introspection test. Some test were actually failing because it was missing. But otherwise I don't know how I could meaningfully test it further. If you have any suggestions I'm happy to add them.

Comment (AndrewSisley) [reply tests/integration/schema/default_fields.go:161]: I actually forgot about the field, I am more concerned about it's params.

You need to test that the InputArgs are properly formed, and only present on fields of the relevant types.

The test directory is a bit of a mess, but `TestFilterForSimpleSchema` is an example of this for filters, `TestSchemaAggregateSimpleCreatesUsersAverage` does the same for `_avg`.

Comment (fredcarle) [reply tests/integration/schema/default_fields.go:161]: Let me know what you think of the added introspection tests.

File: tests/integration/schema/similarity_test.go [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+										"type": map[string]any{
+											"inputFields": []any{
+												map[string]any{
+													"name": "vector",
+													"type": map[string]any{
+														"kind": "NON_NULL",
+														"name": any(nil),
+														"ofType": map[string]any{
+															"kind": "LIST",
+															"name": any(nil),
```
Comment (AndrewSisley) [inline tests/integration/schema/similarity_test.go:83]: suggestion: `"name": nil,` should work fine

File: tests/integration/schema/default_fields.go (hunk 9 above)
Comment (AndrewSisley) [reply tests/integration/schema/default_fields.go:161]: They look good, thanks Fred :)

File: tests/integration/schema/similarity_test.go (hunk 10 above)
Comment (fredcarle) [reply tests/integration/schema/similarity_test.go]: Indeed. Changed.

**Comment by shahzadlone** [comment]
Bug bash + PR Review: I am leaving/ moving this back to the bug bash Todo because it can probably benefit from being tested by one more person, my testing might have missed some other edge cases:

However I noticed a minor bug that was introduced in the debug explain testing framework: https://github.com/sourcenetwork/defradb/issues/3471

Should be very easy to fix and add a debug explain test for it, explained it all in the issue.

**Comment by islamaliev** [comment]
Bug bash result:
After manually updating the embedding field for one of the document I got this message:
"source and vector must be of the same length. Source: 3, Vector: 768"
This is expected, but if there are many document, it might be hard to find one, especially considering that we don't have a way of querying docs with array of certain length. So I would suggest to also return the doc's docID.

After deleting the document querying works as normal.
